src/hoc/withTheme.js

```diff
@@ -33,45 +39,53 @@ const wrapWithTheme = (Component: ReactClass<any>) => {
       [CHANNEL_NEXT]: CONTEXT_CHANNEL_SHAPE,
     }
 
-    state: { theme?: ?Object } = {}
+    static getDerivedStateFromProps(
+      nextProps: {
+        theme?: ?Object,
+        [key: string]: any,
+      },
+      prevState: State
+    ) {
+      const theme = determineTheme(
+        nextProps,
+        prevState.theme,
+        prevState.defaultProps
+      )
+
+      return { theme }
+    }
+
+    styledContext = this.context[CHANNEL_NEXT]
+
+    state: State = {
+      theme:
+        this.styledContext !== undefined
+          ? this.styledContext.getTheme()
+          : undefined,
+      // We need it in state so that gDSFP can access it
+      defaultProps: this.constructor.defaultProps,
```

Comment:
Why does a component need to read its own default props?

Reply:
I’d suspect it’s to determine which theme to use... but it could really be extracted there maybe?

Reply:
The reason I'm confused is that React already takes `defaultProps` into account (and copies them into `props` on element creation). It’s odd for a component to have to “think about” default props explicitly.

Reply:
(I hope I’m going to get this right) the most important theme for us is one that’s passed in as a prop, the “theme” prop. This is so that themes can be overridden for a specific component (rarely used and barely documented though). The standard is that we use the theme from the themeprovider i.e. context however. But if that doesn’t exist we’d like to fall back to the default props. Hence we’re comparing the theme prop to the defaultProps’ theme prop

Reply:
AFAIK styled-components wants to enforce "explicit props > context > defaultProps" priority, there's a comment about it somewhere

Reply:
Yep, that’s the one I was thinking of! Thanks for linking it

Reply:
Eh. Not a fan. From React's point of view default props *are* props, and an optimizing compiler would feel free to "inline" default values at the call site.

Reply:
I can see why you’d want it to work like this though.

Reply:
We can discuss it separately maybe. I never liked the complexity it introduces and “just happened” more organically than consciently. It’s giving us some trouble in GlobalStyles since there the contributor wanted to reimplement it as a component, since there’s no generic “StyledComponent” abstraction yet. The logic proved unintuitive code-wise